### ISSUE TITLE
Could use better time synchronization for game API

### ISSUE THREAD

**Issue by cdsmith** [google/codeworld#431]
Currently, game clients are synchronized relative to the moment they receive the "game started" message from the game server.  If that one packet is delayed (by a TCP drop and retransmit, for example), a client will persistently remain behind the others in time.  That means other clients will appear jerky, as they keep going back to add events in the more distant past, then replaying their own events from that point.  In fairness terms, the lagged client gets an unfair advantage, as it receives other players' events in the future before they actually occur, but other clients don't see its own events until well after they occur.

To fix this, we should add some kind of NTP-like time management code to revise each client's current timestamp toward a consistent value regardless of any original discrepancy.  (But keep in mind that it's critical for correctness that time is monotonic, so we can never decrease a client's timestamp; only slow its rate of increase to allow other clients to catch up.)

**Comment by nomeata** [comment]
> In fairness terms, the lagged client gets an unfair advantage, as it receives other players' events in the future before they actually occur

This is not too bad. “Future” events will not be applied by the client. So the advantage he gets is that he seems them without any lag, but the user cannot “pre-react” to an event.

**Comment by cdsmith** [comment]
Yes, to be clear, the fairness isn't REALLY what I'm worried about.  It's the way that, on one client, the other player's movements all seem to jump forward at the start, then jump back when they finish.

This isn't rare: it happens all the time, even on a fast network.  Maybe some browser was garbage-collecting at the wrong time, or a packet was dropped and retransmitted, or whatever.  All it takes is one delay, and the whole game is shot to hell.  May as well drop the connection and start over.

So this is definitely a priority to fix.

**Comment by nomeata** [comment]
I haven’t observed this yet.
>  It's the way that, on one client, the other player's movements all seem to jump forward at the start, then jump back when they finish.
Is this the a client that starts late, but otherwise receives packages sufficiently fast? If so, then there is a bug. Given our code, would expect that  from the late starting clients p.o.v., all other player’s movements are completely smooth, since all their input has reached this client in time.

I don’t quite see how movements can jump forward and then backwards.
